mackerel-plugin-docker/docker.go

```diff
@@ -0,0 +1,234 @@
+package main
+
+import (
+	"bytes"
+	"errors"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"regexp"
+	"strconv"
+	"strings"
+
+	mp "github.com/mackerelio/go-mackerel-plugin-helper"
+)
+
+var graphdef map[string](mp.Graphs) = map[string](mp.Graphs){
+	"docker.cpuacct.#": mp.Graphs{
+		Label: "Docker CPU",
+		Unit:  "integer",
+		Metrics: [](mp.Metrics){
+			mp.Metrics{Name: "user", Label: "User", Diff: true, Stacked: true, Type: "uint64"},
+			mp.Metrics{Name: "system", Label: "System", Diff: true, Stacked: true, Type: "uint64"},
+		},
+	},
+	"docker.memory.#": mp.Graphs{
+		Label: "Docker Memory",
+		Unit:  "integer",
+		Metrics: [](mp.Metrics){
+			mp.Metrics{Name: "cache", Label: "Cache", Diff: false, Stacked: true},
+			mp.Metrics{Name: "rss", Label: "RSS", Diff: false, Stacked: true},
+		},
+	},
+	"docker.blkio.io_queued.#": mp.Graphs{
+		Label: "Docker BlkIO Queued",
+		Unit:  "integer",
+		Metrics: [](mp.Metrics){
+			mp.Metrics{Name: "read", Label: "Read", Diff: false, Stacked: true},
+			mp.Metrics{Name: "write", Label: "Write", Diff: false, Stacked: true},
+			mp.Metrics{Name: "sync", Label: "Sync", Diff: false, Stacked: true},
+			mp.Metrics{Name: "async", Label: "Async", Diff: false, Stacked: true},
+		},
+	},
+	"docker.blkio.io_serviced.#": mp.Graphs{
+		Label: "Docker BlkIO IOPS",
+		Unit:  "integer",
+		Metrics: [](mp.Metrics){
+			mp.Metrics{Name: "read", Label: "Read", Diff: true, Stacked: true, Type: "uint64"},
+			mp.Metrics{Name: "write", Label: "Write", Diff: true, Stacked: true, Type: "uint64"},
+			mp.Metrics{Name: "sync", Label: "Sync", Diff: true, Stacked: true, Type: "uint64"},
+			mp.Metrics{Name: "async", Label: "Async", Diff: true, Stacked: true, Type: "uint64"},
+		},
+	},
+	"docker.blkio.io_service_bytes.#": mp.Graphs{
+		Label: "Docker BlkIO Bytes",
+		Unit:  "integer",
+		Metrics: [](mp.Metrics){
+			mp.Metrics{Name: "read", Label: "Read", Diff: true, Stacked: true, Type: "uint64"},
+			mp.Metrics{Name: "write", Label: "Write", Diff: true, Stacked: true, Type: "uint64"},
+			mp.Metrics{Name: "sync", Label: "Sync", Diff: true, Stacked: true, Type: "uint64"},
+			mp.Metrics{Name: "async", Label: "Async", Diff: true, Stacked: true, Type: "uint64"},
+		},
+	},
+}
+
+type DockerPlugin struct {
+	Host          string
+	DockerCommand string
+	Tempfile      string
+}
+
+func getFile(path string) (string, error) {
+	cmd := exec.Command("cat", path)
+	var out bytes.Buffer
+	cmd.Stdout = &out
+	err := cmd.Run()
+	if err != nil {
+		return "", err
+	}
+	return out.String(), nil
+}
+
+func exists(path string) (bool, error) {
+	_, err := os.Stat(path)
+	if err == nil {
+		return true, nil
+	}
+	if os.IsNotExist(err) {
+		return false, nil
+	}
+	return true, err
+}
+
+func normalizeMetricName(str string) string {
```

Comment:
You may simplify the code as following.

```
var re = regexp.MustCompile(`[^-a-zA-Z0-9_]`)
re.ReplaceAllString(str, "_")
```

And `regexp.MustCompile` should be located outside of the func.

ref. http://ukai-go-talks.appspot.com/2014/gocon.slide#9